pkg/plugin/framework_methods.go

```diff
@@ -0,0 +1,610 @@
+			"Mismatched SchedulerName for pod: our config has %q, but the pod has %q",
+			e.state.config.SchedulerName, pod.Spec.SchedulerName,
+		)
+		logger.Error("Pod has unexpected SchedulerName", zap.Error(err))
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	return nil
+}
+
+// PostFilter is used by us for metrics on filter cycles that reject a Pod by filtering out all
+// applicable nodes.
+//
+// Quoting the docs for PostFilter:
+//
+// > These plugins are called after Filter phase, but only when no feasible nodes were found for the
+// > pod.
+//
+// PostFilter implements framework.PostFilterPlugin.
+func (e *AutoscaleEnforcer) PostFilter(
+	ctx context.Context,
+	state *framework.CycleState,
+	pod *corev1.Pod,
+	filteredNodeStatusMap framework.NodeToStatusMap,
+) (_ *framework.PostFilterResult, status *framework.Status) {
+	ignored := e.state.config.ignoredNamespace(pod.Namespace)
+
+	e.metrics.IncMethodCall("PostFilter", pod, ignored)
+	defer func() {
+		e.metrics.IncFailIfnotSuccess("PostFilter", pod, ignored, status)
+	}()
+
+	logger := e.logger.With(
+		zap.String("method", "PostFilter"),
+		reconcile.ObjectMetaLogField("FilterPod", pod),
+	)
+	logger.Error("Pod rejected by all Filter method calls")
+
+	return nil, nil // PostFilterResult is optional, nil Status is success.
+}
+
+// Filter gives our plugin a chance to signal that a pod shouldn't be put onto a particular node
+//
+// Filter implements framework.FilterPlugin.
+func (e *AutoscaleEnforcer) Filter(
+	ctx context.Context,
+	_state *framework.CycleState,
+	pod *corev1.Pod,
+	nodeInfo *framework.NodeInfo,
+) (status *framework.Status) {
+	ignored := e.state.config.ignoredNamespace(pod.Namespace)
+
+	e.metrics.IncMethodCall("Filter", pod, ignored)
+	defer func() {
+		e.metrics.IncFailIfnotSuccess("Filter", pod, ignored, status)
+	}()
+
+	nodeName := nodeInfo.Node().Name
+
+	logger := e.logger.With(
+		zap.String("method", "Filter"),
+		reconcile.ObjectMetaLogField("FilterPod", pod),
+		reconcile.ObjectMetaLogField("Node", nodeInfo.Node()),
+	)
+
+	logger.Info("Handling Filter request")
+
+	if status := e.checkSchedulerName(logger, pod); status != nil {
+		return status
+	}
+
+	podState, err := state.PodStateFromK8sObj(pod)
+	if err != nil {
+		msg := "Error extracting local information for Pod"
+		logger.Error(msg, zap.Error(err))
+		return framework.NewStatus(
+			framework.UnschedulableAndUnresolvable,
+			fmt.Sprintf("%s: %s", msg, err.Error()),
+		)
+	}
+
+	// precreate a map for the pods that are proposed to exist on this node, so that we're not doing
+	// this with the lock acquired.
+	proposedPods := make(map[types.UID]*framework.PodInfo)
+	for _, p := range nodeInfo.Pods {
+		proposedPods[p.Pod.UID] = p
+	}
+
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+
+	ns, ok := e.state.nodes[nodeName]
+	if !ok {
+		msg := "Node not found in local state"
+		logger.Error(msg)
+		return framework.NewStatus(framework.Error, fmt.Sprintf("%s: %s", msg, err.Error()))
```

Comment:
issue (blocking): `err` is incorrectly used and is always `nil` here

suggestion:
```suggestion
		return framework.NewStatus(framework.Error, msg)
```

Reply:
Good catch, thanks! Fixed in 3a9f02cb6ce1aecb1193f9ba31cb00ad8aa085af